controllers/reconcilers/kserve_inferenceservice_reconciler.go

```diff
@@ -0,0 +1,171 @@
+
+	//  Resource created per namespace
+	log.V(1).Info("Verifying that the default ServiceMeshMemberRoll has the target namespace")
+	if err := r.istioSMMRReconciler.Reconcile(ctx, log, isvc); err != nil {
+		return err
+	}
+
+	log.V(1).Info("Verifying that the role binding to enable prometheus access exists")
+	if err := r.prometheusRoleBindingReconciler.Reconcile(ctx, log, isvc); err != nil {
+		return err
+	}
+
+	log.V(1).Info("Creating Istio Telemetry object for target namespace")
+	if err := r.istioTelemetryReconciler.Reconcile(ctx, log, isvc); err != nil {
+		return err
+	}
+
+	log.V(1).Info("Creating Istio ServiceMonitor for target namespace")
+	if err := r.istioServiceMonitorReconciler.Reconcile(ctx, log, isvc); err != nil {
+		return err
+	}
+
+	log.V(1).Info("Creating Istio PodMonitor for target namespace")
+	if err := r.istioPodMonitorReconciler.Reconcile(ctx, log, isvc); err != nil {
+		return err
+	}
+
+	log.V(1).Info("Reconciling PeerAuthentication for target namespace")
+	if err := r.istioPeerAuthenticationReconciler.Reconcile(ctx, log, isvc); err != nil {
+		return err
+	}
+
+	log.V(1).Info("Reconciling NetworkPolicy for target namespace")
+	if err := r.networkPolicyReconciler.Reconcile(ctx, log, isvc); err != nil {
+		return err
+	}
+
+	//  Resource created for each ISVC resource
+	log.V(1).Info("Reconciling Generic Route for Kserve InferenceService")
+	if err := r.routeReconciler.Reconcile(ctx, log, isvc); err != nil {
+		return err
+	}
+
+	log.V(1).Info("Reconciling Metrics Service for InferenceService")
+	if err := r.metricsServiceReconciler.Reconcile(ctx, log, isvc); err != nil {
+		return err
+	}
+
+	log.V(1).Info("Reconciling Metrics ServiceMonitor for InferenceService")
+	if err := r.metricsServiceMonitorReconciler.Reconcile(ctx, log, isvc); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *KserveInferenceServiceReconciler) OnDeletionOfKserveInferenceService(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Deleting Kserve inference service generic route")
+	return r.routeReconciler.DeleteRoute(ctx, isvc)
+}
+
+func (r *KserveInferenceServiceReconciler) DeleteKserveMetricsResourcesIfNoKserveIsvcExists(ctx context.Context, log logr.Logger, isvcNamespace string) error {
+	inferenceServiceList := &kservev1beta1.InferenceServiceList{}
+	if err := r.client.List(ctx, inferenceServiceList, client.InNamespace(isvcNamespace)); err != nil {
+		return err
+	}
+
+	for i := len(inferenceServiceList.Items) - 1; i >= 0; i-- {
+		inferenceService := inferenceServiceList.Items[i]
+		if utils.IsDeploymentModeForIsvcModelMesh(&inferenceService) {
+			inferenceServiceList.Items = append(inferenceServiceList.Items[:i], inferenceServiceList.Items[i+1:]...)
+		}
+	}
+
+	// If there are no Kserve InferenceServices in the namespace, delete namespace-scoped resources needed for Kserve Metrics
+	if len(inferenceServiceList.Items) == 0 {
```

Comment:
why can't re rely on the owner metadata information to delete everything?

Reply:
Some of the resources are not directly dependent on instance of InferenceResource.For example IstioSMMR, PrometheusRoleBinding, IstioTelemetry, IstioServiceMonitor, IstioPodMonitor, IstioPeerAuthentication and NetworkPolicy resources are supposed to be created only once for a namespace that why we didn't set the Owner references to them and only delete them when all InferenceResource resources are deleted from that namespace.

Reply:
I see, thanks for the explanation.